core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -107,13 +107,18 @@ private[log] class ProducerAppendInfo(val producerId: Long,
         throw new DuplicateSequenceNumberException(s"Duplicate sequence number for producerId $producerId: (incomingBatch.firstSeq, " +
           s"incomingBatch.lastSeq): ($firstSeq, $lastSeq), (lastEntry.firstSeq, lastEntry.lastSeq): " +
           s"(${this.firstSeq}, ${this.lastSeq}).")
-      } else if (firstSeq != this.lastSeq + 1L) {
+      } else if (firstSeq != this.lastSeq + 1L && (firstSeq != 0 && this.lastSeq != Short.MaxValue)) {
         throw new OutOfOrderSequenceException(s"Out of order sequence number for producerId $producerId: $firstSeq " +
           s"(incoming seq. number), ${this.lastSeq} (current end sequence number)")
       }
     }
   }
 
+  private def isProducerFenced(producerEpoch: Short): Boolean = {
+      producerEpoch < 0 || (producerEpoch < this.producerEpoch &&
```

Comment:
I think this matches existing logic: `NO_EPOCH` is guaranteed to be fenced (though really it should be an invalid request).